internal/envoy/install.go

```diff
@@ -75,9 +79,68 @@ func InstallIfNeeded(ctx context.Context, o *globals.GlobalOpts, v version.Versi
 		// TODO: figure out how to get a stat error that isn't file not exist so we can test this
 		return "", err
 	}
+
+	// Set the inferred Envoy version here, hence it can be used in the subsequent steps.
+	o.EnvoyVersion = v
+
 	return verifyEnvoy(installPath)
 }
 
+func ensureVersionWithPatch(ctx context.Context, o *globals.GlobalOpts, v version.Version) (version.Version, version.ReleaseVersions, error) {
+	var ev version.ReleaseVersions
+	var err error
+
+	// When the given version has no patch component, we fetch all available versions and check if
+	// there is a compatible versions available.
+	if o.EnvoyVersionHasNoPatch {
+		ev, err = FuncEVersions(ctx, o.EnvoyVersionsURL, o.Platform, v)
+		if err != nil {
+			return "", ev, err
+		}
+		compatibleVersions := ev.CompatibleVersionStrings()
```

Comment:
both the logic here and CompatibleVersionStrings and its corresponding tests feel like too much code for the job. Also it is looser than it needs to be. There's only one call site as far as I can tell (here), so if you define logic here until there are multiple call sites. That way the function can be un exported and focus only on getting the thing.. iterative seems simplest. I think this can reduce the lines and things to explain, test, review during future refactorings as FuncEVersions can be left alone.

```
// findLatestVersion returns the latest available patch or empty if not found
findLatestVersion(ev version.ReleaseVersions, minor string) string {
var latestPatch
for _, patch ... {
  if patch has prefix minor && patch > latestPatch {
latestPath = patch
}
}
return latestPatch
```